### PR TITLE
CDRIVER-4352 unskip `/Samples` tests

### PR DESCRIPTION
**Pull request by kevinAlbs**
# Summary

Resolves CDRIVER-4352. Unskips the `/Samples*` tests and applies fixes.

To verify changes, the `/Samples*` tests were run 100 times in [this patch build](https://spruce.mongodb.com/version/66b41bb6f1658d0007fc2a18).

# Details

Several test errors were addressed. See commits for details. Notably:

- An informational `MONGOC_INFO` log was replaced with `MONGOC_DEBUG`. This was to fix a failing `ASSERT_NO_CAPTURED_LOGS`. CDRIVER-4282 changed test log capture to [include `MONGOC_LOG_LEVEL_INFO`](https://github.com/mongodb/mongo-c-driver/pull/1374/files#diff-b71f20ac0b55a16f43592d6bc46791721a0a075febda9e508f3c66d555fae9e9L209). 
- Snapshot query examples added in CDRIVER-4295 were updated to skip on server <= 5.0. Drivers [error](https://jira.mongodb.org/browse/DRIVERS-1820) if using snapshot sessions with server < 5.0.
- Snapshot query examples now use majority write concern to insert data. This is to address errors observed in Evergreen [described here](https://jira.mongodb.org/browse/DRIVERS-2181?focusedCommentId=6589519&focusedId=6589519&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-6589519).

Drive-by improvements:

- Skip `test_sample_causal_consistency` on single-node replica sets. The test requires a secondary. This is not strictly needed (as Evergreen tests with three-node replica sets) but is intended as a local convenience. This can be removed if preferred.
- `capture_logs (false)` calls were added to reduce scope of log capture.
- The snapshot query examples were renamed to avoid using the numbers 59 and 60. The numbers 1-58 are associated with [examples in DRIVERS-356](https://jira.mongodb.org/browse/DRIVERS-356?focusedCommentId=1521532&focusedId=1521532&page=com.atlassian.jira.plugin.system.). 59 and 60 have no known reference.

### CODE REVIEW

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 1]
```diff
@@ -53,6 +53,8 @@ test_sample_command (sample_command_fn_t fn,
 
    ASSERT_NO_CAPTURED_LOGS (example_name);
 
+   capture_logs (false);
```
Comment (eramongodb) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:56]: Suggest moving up further to wrap the call to `fn`:

```c
capture_logs (true);
fn (db);
capture_logs (false);
ASSERT_NO_CAPTURED_LOGS (example_name);
```

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 2]
```diff
@@ -2590,18 +2602,18 @@ accumulate_adoptable_count (const mongoc_client_session_t *cs,
...
-test_example_59 (mongoc_database_t *db)
+test_snapshot_query_example_1 (void)
 {
-   BSON_UNUSED (db);
-
    if (!test_framework_skip_if_no_txns ()) {
       return;
    }
 
+   if (test_framework_get_server_version () < test_framework_str_to_version ("5.0.0")) {
```
Comment (eramongodb) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:2612]: Can/should this use wire version 13 (`WIRE_VERSION_5_0`) instead?

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 3]
```diff
@@ -2822,6 +2847,18 @@ test_example_60 (mongoc_database_t *db)
...
+// `test_snapshot_query_examples` examples for DRIVERS-2181.
+static void
+test_snapshot_query_examples (void)
+{
+   capture_logs (true);
+   test_snapshot_query_example_1 ();
+   ASSERT_NO_CAPTURED_LOGS ("test_snapshot_query_example_1");
+   test_snapshot_query_example_2 ();
+   ASSERT_NO_CAPTURED_LOGS ("test_snapshot_query_example_2");
+   capture_logs (false);
```
Comment (eramongodb) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:2859]:
```suggestion
   capture_logs (true);
   test_snapshot_query_example_1 ();
   capture_logs (false);
   ASSERT_NO_CAPTURED_LOGS ("test_snapshot_query_example_1");

   capture_logs (true);
   test_snapshot_query_example_2 ();
   capture_logs (false);
   ASSERT_NO_CAPTURED_LOGS ("test_snapshot_query_example_2");
```
Recommend consistent pattern of minimizing capture "scope" as much as possible.

Also applies to other instances of `capture_logs (false)` added in this PR.

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 4]
```diff
@@ -3018,6 +3056,11 @@ test_sample_causal_consistency (mongoc_client_t *client)
       return;
    }
 
+   if (test_framework_is_replset () && test_framework_data_nodes_count () == 1) {
```
Comment (eramongodb) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:3059]: Should this condition be `!test_framework_is_replset () || test_framework_data_nodes_count () == 1)`? Does this test support running against a sharded cluster?

**Review by eramongodb — CHANGES_REQUESTED** [review]

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 5]
```diff
@@ -2609,8 +2607,8 @@ test_snapshot_query_example_1 (void)
       return;
    }
 
-   if (test_framework_get_server_version () < test_framework_str_to_version ("5.0.0")) {
-      MONGOC_DEBUG ("Skipping test. Test requires server version 5.0.0\n");
+   if (!test_framework_max_wire_version_at_least (MONGOC_READ_SECONDARY)) {
```
Comment (eramongodb) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:2610]:
```suggestion
   if (!test_framework_max_wire_version_at_least (WIRE_VERSION_SNAPSHOT_READS)) {
```
To match the updated message below. (`MONGOC_READ_SECONDARY` is an unrelated macro constant.)

**Review by eramongodb — APPROVED** [review]
One last fix; otherwise, LGTM.

File: src/libmongoc/tests/test-mongoc-sample-commands.c (hunk 5 above)
Comment (kevinAlbs) [reply src/libmongoc/tests/test-mongoc-sample-commands.c:2610]: Oof 🤦. Well spotted.

**Pull request merged**